### ISSUE TITLE
Avoiding plan and execute for stopped vehicles

### ISSUE THREAD

**Issue by behrisch** [eclipse/sumo#8110]
This is an idea for simulation speedup. Vehicles which are on a stop still do the whole plan/execute cycle with a foreseeable result. Optimizations could take involve three steps:

- [ ] Skip calculations for vehicles which have at least one more step at their current stop (maybe using the actionstep mechanism)
- [ ] Skip calculations for vehicles which wait at a traffic light which will not turn green in the next step
- [ ] skip it for vehicles waiting for another vehicle fulfilling one of these conditions (transitively)

**Comment by namdre** [comment]
> Skip calculations for vehicles which have at least one more step at their current stop (maybe using the actionstep mechanism)

Vehicles that must stop abort the plan move loop early: https://github.com/eclipse/sumo/blob/16ba97d7de80a2defc7061834d3c30feb9094368/src/microsim/MSVehicle.cpp#L2199-L2201

> Skip calculations for vehicles which wait at a traffic light which will not turn green in the next step

The traffic light for the current step switches ahead of the planMove step. Vehicles already abort their planning quickly once they encounter linkstate red: https://github.com/eclipse/sumo/blob/16ba97d7de80a2defc7061834d3c30feb9094368/src/microsim/MSVehicle.cpp#L2401-L2413. 

> skip it for vehicles waiting for another vehicle fulfilling one of these conditions (transitively)

Somewhat similar to my comment above: the vehicles might still advance toward their leader. After that has been checked, they will hit the same red link and abort. Also, if the speed drops to 0 in the planning step, the loop is aborted as well:
https://github.com/eclipse/sumo/blob/16ba97d7de80a2defc7061834d3c30feb9094368/src/microsim/MSVehicle.cpp#L2552-L2554

In summary, don't think there is much to optimize here.

**Comment by behrisch** [comment]
Good to hear! I had a different impression stepping through the code on the collision stop issue but I will recheck.

**Comment by behrisch** [comment]
> Vehicles that must stop abort the plan move loop early:
> 
> https://github.com/eclipse/sumo/blob/16ba97d7de80a2defc7061834d3c30feb9094368/src/microsim/MSVehicle.cpp#L2199-L2201
> 

This does not look to me as if this break saves much here. Yes it doesn't look into further lanes, but there are still a lot of calculations after the loop. Do you have a good idea how we should measure this or should we just add stopwatches again?

**Comment by namdre** [comment]
> This does not look to me as if this break saves much here. Yes it doesn't look into further lanes, but there are still a lot of calculations after the loop. Do you have a good idea how we should measure this or should we just add stopwatches again?

My guess would be that the potential savings in executeMove are small since there is only one driveProzessItem (with setRequest=false) and the percentage of stopped vehicles in any given simulation step is small anyway. Haling vehicles on the other hand could represent the major share in a jammed simulation.

Making processLaneAdvances and updateFurtherLanes conditional on having moved some distance might help.

Also, having a fully jammed simulation (i.e. by keeping all traffic lights red) might make for a good gprof test case.
